structs-methods-interfaces/shapes_test.go

```diff
@@ -0,0 +1,37 @@
+package main
+
+import "testing"
+
+func TestPerimeter(t *testing.T) {
+	rectangle := Rectangle{10.0, 10.0}
+	got := Perimeter(rectangle)
+	want := 40.0
+
+	if got != want {
+		t.Errorf("got %.2f want %.2f", got, want)
+	}
+}
+
+func checkArea(t testing.TB, shape Shape, want float64) {
+	t.Helper()
+	got := shape.Area()
+	if got != want {
+		t.Errorf("got %g want %g", got, want)
+	}
+}
+func TestArea(t *testing.T) {
+
+	areaCases := []struct {
+		shape Shape
+		want  float64
+	}{
+		{shape: Rectangle{Width: 12, Height: 6}, want: 72.0},
+		{shape: Circle{Radius: 10}, want: 314.1592653589793},
+		{shape: Triangle{Base: 12, Height: 6}, want: 36.0},
+	}
+
+	for _, c := range areaCases {
+		checkArea(t, c.shape, c.want)
+	}
+
+}
```

Comment:
Where are the examples and benchmarks? :-D

Reply:
I know. Will add them today :)